(fix) O3-4131: Improve Transition patient and Edit modals

harshthakkr opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary
This PR introduces the following changes to the Transition patient & Edit modals:
1. Renamed the action button from "Transition" to "Move" in Transition modal
2. Renamed the modal title from "Transition patient" to "Move {patient}" in Transition modal
3. If locations are <= 8 then radio buttons are displayed otherwise a dropdown with white background is displayed
4. Modal width is set to 480px
5. Time of transition option is completely removed
6. Used radio buttons with priority tags to show the Queue priority
7. Priority comment is only available if the priority is set to "Urgent" or "Emergency"
8. If user sets priority to "Not urgent" then previously stored Priority comment will be removed

## Screenshots
### Case 1: Locations are > 8 and Priority comment feature

https://github.com/user-attachments/assets/a784b202-14a4-4d7f-ba58-8454da4a2c78

### Case 2: Locations are <= 8
<img width="1470" alt="Screenshot 2025-02-21 at 1 13 18 AM" src="https://github.com/user-attachments/assets/f631efb6-10db-4e36-892a-d1dea5631b2a" />

### Case 3: Edit modal
<img width="1470" alt="Screenshot 2025-02-25 at 9 50 17 AM" src="https://github.com/user-attachments/assets/7de0c268-4981-4a8c-b4ec-5afa39300dcd" />

## Related Issue
[O3-4131](https://openmrs.atlassian.net/jira/software/c/projects/O3/issues/O3-4131?filter=myopenissues&jql=project%20%3D%20%22O3%22%20AND%20assignee%20IN%20%28currentUser%28%29%29%20AND%20statusCategory%20IN%20%28%22To%20Do%22%2C%20%22In%20Progress%22%29%20ORDER%20BY%20created%20DESC)

## Note
The updates to the Transition string are applied only to the English language (i.e., en.json), as I am not familiar with other languages.

denniskigen commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -214,33 +153,60 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
-                  />
-                ))}
-              </Select>
+              {queues.length <= 8 ? (
```
```suggestion
              {queues.length <== 8 ? (
```

harshthakkr replied:
I think `<==` operator doesn't exist in JS.

denniskigen replied:
Oops

denniskigen replied:
Can we add a comment explaining why we chose to use 8 here?

harshthakkr replied:
Ah, the hardest part of development - writing a good comment! 😅

`{queues.length <= 8 ? (
// here value 8 is used as mentioned in the issue #O3-4131 by Ciarán Duffy
<RadioButtonGroup`

Does this look fine to you, @denniskigen, or would you suggest any changes?

harshthakkr replied:
Thanks!

denniskigen commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -287,89 +252,53 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
+                      name={display}
+                      labelText={
+                        <Tag
+                          className={`${priorities.findIndex((p) => p.uuid === uuid) === 1 && styles.orange} ${
```
It looks like we're doing this array index check both here and in the `type` property below. Can we centralize this logic into a function? Also, consider using `classnames` for merging multiple classes conditionally.

denniskigen commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -287,89 +252,53 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
+                          {uuid == queueEntry.priority.uuid
+                            ? t('currentValueFormatted', '{{value}} (Current)', {
+                                value: display,
+                                interpolation: { escapeValue: false },
```
Using `escapeValue` opts us out of `i18next` input sanitization and means we have to roll our own mechanism to mitigate against potential XSS attacks. See https://github.com/openmrs/openmrs-esm-patient-chart/pull/2230 for a possible alternative solution.

denniskigen commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -214,33 +153,60 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
+                        uuid == queueEntry.queue.uuid
+                          ? t('currentValueFormatted', '{{value}} (Current)', {
+                              value: `${display} - ${location?.display}`,
+                              interpolation: { escapeValue: false },
```
Ditto here about [unescaping](https://www.i18next.com/translation-function/interpolation#unescape) input values.

denniskigen reviewed (approved):
Good start, @harshthakkr. I've highlighted a few things we should address.

denniskigen reviewed (commented):
Good start, @harshthakkr. I've highlighted a few things we should address.

denniskigen reviewed (commented):
Good start, @harshthakkr. I've highlighted a few things we should address.

harshthakkr commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -172,40 +145,6 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
-      : null;
-
-    if (startAtDate > now) {
-      return t('timeCannotBeInFuture', 'Time cannot be in the future');
```
Shouldn't we remove unused translation strings from all locale JSON files?

harshthakkr replied:
> Shouldn't we remove unused translation strings from all locale JSON files?

Could you clarify this, @denniskigen?

denniskigen replied:
> Shouldn't we remove unused translation strings from all locale JSON files?

No, we should only remove them from the `en.json` file. The Transifex integration will handle purging any unused key/value translation string pairs automatically.

harshthakkr commented:
@denniskigen could you please review it?

denniskigen reviewed (commented):
LGTM. And should be good to go after we finish the release cuts.

chibongho commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -287,89 +251,52 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
+                          value={uuid}
+                          type={(() => {
+                            const index = findPriorityIndex(uuid);
+                            return index === 0 ? 'green' : index === 2 ? 'red' : '';
```
```suggestion
                            // TODO: fix priority colors. https://openmrs.atlassian.net/browse/O3-4469
                            return index === 0 ? 'green' : index === 2 ? 'red' : '';
```

We can keep this for now, since we want to change it anyway. But... This assumes that the queue has 3 priorities in order of least urgent to most. This works for the queues in the refapp, but is not a safe assumption in general.

chibongho commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -287,89 +251,52 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
-                </TimePicker>
-              </div>
-            </section>
+            {selectedPriorityIndex > 0 && (
```
We might want to keep this. Let's wait for discussion in the ticket. (If we do want to hide this conditionally based on the priority, it will need to be configurable.)

chibongho commented on packages/esm-service-queues-app/src/queue-table/queue-entry-actions/queue-entry-actions.modal.tsx:
```diff
@@ -287,89 +251,52 @@ export const QueueEntryActionModal: React.FC<QueueEntryActionModalProps> = ({
               )}
             </section>
-
-            <section className={styles.section}>
```
Per @fanderson's comment in the ticket, we will likely need to keep this. My guess is that we'll add a config value to show/hide this, but let's wait for discussion in the ticket.

chibongho reviewed (commented):
Sorry, some of the behavioral changes mentioned in the ticket should have required more discussion before implementing. Otherwise, the UI changes look good.

harshthakkr commented:
I accidentally created this commit mess while resolving merge conflicts, extremely sorry about that. Updated the PR based on the discussion on JIRA. Let me know if I need to change anything. [Design link](https://app.zeplin.io/project/6650717a9275db731847fb32/screen/67d2fbe66ce773a720ec6d5e)

### Changes made
1. Renamed 'Priority comment' to 'Comment'
2. Comment box will now appear irrespective of the selected priority
3. Added an 'Advanced options' button to change the date and time

https://github.com/user-attachments/assets/df61c3b7-fdb5-48c6-a8ea-9e5b45ed5c3d

harshthakkr commented:
Could you please review the latest changes @chibongho, @denniskigen?

harshthakkr commented:
Could you please review it @chibongho?

harshthakkr commented on packages/esm-service-queues-app/src/queue-table/cells/queue-table-action-cell.component.tsx:
```diff
@@ -32,6 +33,7 @@ export function QueueTableActionCell({ queueEntry }: QueueTableCellComponentProp
             const dispose = showModal('edit-queue-entry-modal', {
               closeModal: () => dispose(),
               queueEntry,
+              size: 'sm',
```
Yes, it does.

harshthakkr commented:
@chibongho, this is how it looks after the final changes:

https://github.com/user-attachments/assets/dd385917-e792-4b2b-a44f-d03b96d47588
